ni_measurementlink_service/measurement/service.py

```diff
@@ -128,18 +130,35 @@ class MeasurementService:
 
     """
 
-    def __init__(self, measurement_info: MeasurementInfo, service_info: ServiceInfo) -> None:
-        """Initialize the Measurement Service object with measurement info and service info.
+    def __init__( self, service_config_path: str, version: str, ui_file_paths: List[str]) -> None:
+        """Initialize the Measurement Service object with the .serviceconfig file, version, and UI file paths.
 
         Args:
         ----
-            measurement_info (MeasurementInfo): Measurement Info
+            service_config_path (str): Measurement Info
 
-            service_info (ServiceInfo): Service Info
+            version (str): Version of the measurement service.
+
+            ui_file_paths (List[str]): Service Info
 
         """
-        self.measurement_info: MeasurementInfo = measurement_info
-        self.service_info: ServiceInfo = service_info
+        if not path.exists(service_config_path):
+            raise Exception(f"File does not exist. {service_config_path}")
+
+        with open(service_config_path) as service_config_file:
+            service_config = json.load(service_config_file)
+
+        self.measurement_info = MeasurementInfo(
+            display_name=service_config["services"][0]["displayName"],
```

Comment:
`.serviceconfig` files can support multiple services. So simply indexing here doesn't seem ideal. As an alternative, we could create a ServiceConfigParser class and then the user can select the particular service they want and pass that to `__init__` instead.

Alternative to that, we could have a parameter on `__init__` that takes the index we want to use. This could even have a default value of 0. This could cover the majority of cases, but still provide a way to specify a particular service. This would come with the cost of potentially parsing the json file multiple times.

Reply:
Some more alternatives.

**Add a JSON pointer/path parameter** 
There's a standard way to specify a JSON sub-object. Two standards, actually: JSON Pointer ("/services/0") and JSON Path ("$.services[0]"). There are 3rd party modules for these. However, I don't think we need the extra flexibility provided by this approach.

**Continue supporting the old API.**
If we had already shipped this module, we would have to keep support for `measurement_info` and `service_info`. This could also be used as a fallback to support more advanced use cases.

**Optionally specify the service class.**
Add `service_class=None`. This is more maintainable than passing an index because you can see whether it's correct without looking at the `.serviceconfig` file and you don't need to shift all of the other service indices by one when you add/remove a service.
- If `service_class is None`, use the 1st service in the `.serviceconfig` file. This could even validate that there is only one service.
- Otherwise, find the service in the `.serviceconfig` file that matches this service class.

I think I like "Optionally specify the service class" best. I'm not too concerned with the overhead of reading/parsing the `.serviceconfig` file multiple times.

Reply:
I think we should continue to support passing `measurement_info` directly like @bkeryan said to
> be used as a fallback to support more advanced use cases.

Reply:
AFAIK you cannot overload `__init__`. We could have it take a generic `args` parameter and then try to figure out whats being passed in, but that seems very error prone.

We could remove the arguments and create a couple of methods `from_measurement_info` or `from_service_config` to initialize them. That allows for someone forgetting to initialize.

If we want to support the current way of creating the object, I think the best solution would be to create a new class that creates the `measurement_info` and `service_info` objects. These can then be passed into `__init__`.

I think overall the best solution is to prevent the user from dual sourcing the information all together by doing what I am doing now with @bkeryan 's suggestion.

Reply:
Yeah, I guess the .serviceconfig approach seems clean. This may be unfounded, but it seems kind of limiting in a broader sense to make them pass a 'path' to an object. I'm thinking about the overall [Configuration](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/configuration/?view=aspnetcore-7.0) approach that ASP.NET Core takes in order to be very flexible about these kinds of configuration. Obviously that's overkill for what we need, but maybe there are elements of the hierarchy of configuration methods that we could have.

Anyway, I'm ok with going ahead with the .serviceconfig approach, but we either need to make this change quickly before 1.0 or allow both ways and add this after 1.0.

Reply:
> AFAIK you cannot overload __init__. We could have it take a generic args parameter and then try to figure out whats being passed in, but that seems very error prone.

Correct.

My recommendation for supporting both approaches is to make the parameters optional and specify them by name. You can also declare keyword-only arguments by adding a `*` to the parameter list.

```
def __init__( 
        self, 
        *,
        measurement_info: MeasurementInfo = None, 
        service_info: ServiceInfo = None,
        service_config_path: str = "", 
        version: str = "", 
        ui_file_paths: List[str] = []
        ) -> None:

Reply:
> We could remove the arguments and create a couple of methods from_measurement_info or from_service_config to initialize them. That allows for someone forgetting to initialize.

Yeah, you could define a class method `MeasurementInfo.from_service_config(service_config_path)`

Reply:
What happens then if you specify both the objects and the service config file path?

Reply:
> This may be unfounded, but it seems kind of limiting in a broader sense to make them pass a 'path' to an object. I'm thinking about the overall [Configuration](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/configuration/?view=aspnetcore-7.0) approach that ASP.NET Core takes in order to be very flexible about these kinds of configuration.

`*.serviceconfig` and `appsettings.json` have different roles:
- `*.serviceconfig` tells the discovery service when/how to launch the measurement service.
- `appsettings.json` tells the measurement service about itself.

In our C# services, we are storing the `.serviceconfig` filename in `appsettings.json` instead of hardcoding it in the source code. If we had an equivalent configuration file in Python, we could do that here too.

Reply:
> What happens then if you specify both the objects and the service config file path?

Whatever you decide.

Reply:
IMO having both is a bit confusing and complicates the API. Additionally, I don't think this prevents breaking compatibility. In all our examples we pass the `MeasurementInfo` and `ServiceInfo` directly without named parameters. I think we'd have to keep those as positional parameters to prevent breaking changes, but I also feel like that would imply that those parameters are preferred over the other optional parameters.

If we are concerned about backwards compatibility, I can add them back, and we can try to document this as best as possible, but I still think a factory class that takes the `.serviceconfig`, version, and UIs, and returns a `MeasurementInfo` and `ServiceInfo` would be a cleaner approach and maintain compatibility.

Reply:
> IMO having both is a bit confusing and complicates the API.

I agree. If we want to remove the old interface, we should do it before we release 1.0.

> I think we'd have to keep those as positional parameters to prevent breaking changes, but I also feel like that would imply that those parameters are preferred over the other optional parameters.

Right, making the existing parameters keyword-only would break compatibility. 

You could make the existing parameters optional, though.

```
def __init__( 
        self, 
        measurement_info: MeasurementInfo = None, 
        service_info: ServiceInfo = None,
        version: str = "", 
        ui_file_paths: List[str] = []
        *,
        service_config_path: str = "", 
        ) -> None:
```

Reply:
Yeah that would work. I'm not convinced of the usefulness of letting people dual source their measurement/service info, nor why this approach would be better than serviceconfig parsing class.

Reply:
I agree. Better to single source it now.